language/compiler/ir-to-bytecode/syntax/src/lexer.rs

```diff
@@ -114,6 +114,7 @@ impl Tok {
 }
 
 pub struct Lexer<'input> {
+    pub spec_mode: bool,
```

Comment:
I think using a boolean flag to fork the behavior of the lexer in this way is not very robust. For example: what if we want a functional test that runs both the prover and the concrete interpreter? Today we can produce a single AST, compile it to bytecode, and run all of the passes. But with this approach, we'd either need to produce two ASTs (one `spec_mode`, one not) or be unable to write such a test.

 I think it would be better to either:
- Fix the hacky treatment of dots in the current lexer so we can use them in both spec syntax and Move concrete syntax (preferred)
- Leave the slashes in specs, as it's just syntax

Reply:
Sam, I agree that it's probably fragile and generally lacking in taste.  I don't understand your specific point about the interpreter. The flag doesn't depend on whether the prover is used, it is set by the parser when it see "ensures", etc. and cleared when it reaches the end of the specification construct.

I would like to proceed with this pull request so that we have the nicer syntax for specifications, and clean up the parser later.  Arguably, the current lexer and parser are already less tasteful than my hack, and THEY'RE in the repository.

I agree that the best solution would be to  make the lexer's treatment of tokens less weird in general.  I thought about doing that, but it's obviously a bigger project and I wanted to talk to Todd about it, first, since he apparently did the first lexer from which the current code was derived.

Reply:
> That might be a little fragile for other reasons, especially with syntactically incorrect code, but should produce the same AST whenever an MVIR program is parsed for any purpose.

My bad, I misunderstood what the flag is doing. I thought the user had to set it when creating a lexer, but I see now that the code sets and then unsets it. That's not as bad.

> I thought about trying to clean up the treatment of tokens in the existing lexer and parser, but that seems like a potentially bigger job, and I didn't want to do that without talking to Todd, who (according to Bob) came up with the original lexing scheme.

I guess my general feeling is that we should fix existing hacks rather than add new ones even if fixing existing hacks is more work. I think this is particularly true if the new hacks are to support cosmetic changes like different syntax, but I think we may have different opinions on how ugly the slashes are :).

Reply:
We really need to figure out our plan for handling specifications in the source language. In the move-lang compiler, the only special case of punctuation after an identifier is for type arguments (e.g., "identifier<types...>") so there will be no need for this special mode to use dots instead of slashes.

Reply:
I actually think it would be good to move the specification language to the move-lang compiler and abandon the IR. I think the sooner we do it the better it is and we may also have more of a chance to include more people in the specification. 
That is my opinion of course and I may be off a bit ;)

Reply:
Totally agree that specs based on move-lang is the right way to go. But we also have to make progress with actually specifying and proving things and stabilizing the core of the prover logic, which is pretty much independent of the syntax. We have just switched to use the IR based on Sam's initiative, and another refactoring right now would throw us back.

Can we get approval for this change (or any modified change which fixes the strange `a/b` field selection problem) and then work together on a medium term plan (i.e. within the next 2-3 month) to move the spec language to its final home?

Reply:
Personally I don't care what you do with the IR compiler since we expect to replace it soon, but I share Dario's opinion that you should not wait to get started figuring out what to do in move-lang.

Reply:
added few people to this review which may be more "in tune" with the current IR and evolution of it.
I guess changes to the IR should involve those people the most so please make sure to include them. Of course also fine getting things in once they are reviewed by someone in the team in general, I am not trying to slow things down but I guess I am in as good a position as any random person to approve this so I'll delegate ;)
I know others (@tnowacki) may not be quite as ready to get the specification into move-lang, but I also know we are planning to change the IR dramatically (@sunshowers is planning significant work there). So I believe this should be discussed among the interested parties which does not involve me (and yet here I am talking).
As a good disrupter though I have to say I am not a big fun of those little playground we create locking things down (the IR). I understand we need to make progress but it would also be good to see this worked in concert with other efforts. Yes, it will slow things down a bit at the beginning but I am not clear it would be a negative overall, in fact I am convinced it would be a **massive** benefit overall.
Now having said so I also know I may be the only one believing it so I may have to look at the mirror a bit and consider the fact the guy on the other side may be not quite right....

Reply:
@dariorussi I don't think this change has that many implications.
1. The new mvir is not going to be based on this code (if I remember from my conversation with Todd), and there is no reason to put specifications in it.
2. This mvir is going to be discarded soon.
3.  The comments on specs in the Move language are helpful and we'll start thinking about that more seriously, but we don't need to delay this minor change in syntax in order to do that.

Reply:
yes I agree we don't need to delay anything here, this was more about the original comment on starting to think (and implement possibly) the specification language in the context of move lang rather than blocking this PR. 
I have not worked on the IR for a while and that is why I did not accept this and let others with more context comment.
I may have made more a big deal than I intended to, I'll accept soon if no one else does it